versions/development/SPEC.md

```diff
@@ -344,20 +345,77 @@ Tasks and workflow inputs may be passed in from an external source, or they may
-|`\"`|double quote|`\x22`|within a double-quoted string|
+| Escape Sequence | Meaning      | \x Equivalent | Context                       |
+| --------------- | ------------ | ------------- | ----------------------------- |
+| `\\`            | `\`          | `\x5C`        |                               |
+| `\n`            | newline      | `\x0A`        |                               |
+| `\t`            | tab          | `\x09`        |                               |
+| `\'`            | single quote | `\x27`        | within a single-quoted string |
+| `\"`            | double quote | `\x22`        | within a double-quoted string |
 
 Strings can also contain the following types of escape sequences:
 
 * An octal escape code starts with `\`, followed by 3 digits of value 0 through 7 inclusive.
 * A hexadecimal escape code starts with `\x`, followed by 2 hexadecimal digits `0-9a-fA-F`. 
 * A unicode code point starts with `\u` followed by 4 hexadecimal characters or `\U` followed by 8 hexadecimal characters `0-9a-fA-F`.
 
+##### Multi-line Strings
+
+Strings that begin and end with three consecutive single- or double-quotes may span multiple lines. The opening quotes of a multi-line string may be followed by whitespace (optional) and must have a newline before the first non-whitespace character; these leading whitespace/newline characters are removed. All subsequent non-empty lines are then used to determine the multi-line string's *common leading whitespace* - the minimum number of whitespace characters occuring before the first non-whitespace character in a line or the end of the line, whichever comes first. This common leading whitespace is stripped from the beginning of all the lines in the multi-line string.
+
+```wdl
+# These three strings are equivalent. The middle line of each is empty and so does not count
+# towards the common leading whitespace determination.
+
+String multi_line_A = '''
+        this is a
+
+          multi-line string'''
+
+String multi_line_B = """
+  this is a
+
+    multi-line string"""
+
+String multi_line_C = """
+this is a
+
+  multi-line string"""
+```
+
+Newline characters are not stripped out unless they are escaped, i.e. when the last character of a line is `\`.
+
+```wdl
+# These two strings are equivalent:
+String single_line = "this is a double-quoted string that contains no newlines"
+String multi_line_escaped = """
+  this is a \
+  double-quoted string \
+  that contains no newlines"""
+```
+
+Keep in mind that if the ending quotes are on a line by themselves and are preceeded by whitespace, that line is included when determining the common leading whitespace.
+
+```wdl
+# The following two strings are equivalent:
+
+String s1 = "    text indented by 4 spaces"
+
+# Even though the first line of this string is indented by six spaces, the common leading
+# whitespace in this string is 2, due to the two spaces preceeding the closing quotes.
+String s2 = """
+      text indented by 4 spaces
+  """
```

Comment:
IMO, if the last line is whitespace only, it shouldn't count towards the _common leading whitespace_.

Reply:
This is in line with how Java text blocks work.
e.g. see the example under section 3 here https://www.baeldung.com/java-text-blocks

Reply:
But it is not in line with how the command block works in WDL. I predict this is going to lead to questions.
```WDL
String s2 = """
    Text that I do not want indented
    But I do want it indented in the source file
    Indentation in the source makes it easier to read
""" 

 # Looks correct, but leads to the wrong result
```
```WDL
String s2 = """
    Text that I do not want indented
    But I do want it indented in the source file
    Indentation in the source makes it easier to read"""

# Looks wrong, but leads to the correct result.
```

Reply:
You're right. And actually I don't really like this aspect of java text blocks because the handling of the first and last lines feels inconsistent. If we ignore whitespace on the first line we should also ignore it on the last line. That consistency - as well as having consistency between multi-line strings and the command block, seems more important to me then the small benefit of being able to use the last line to "override" the indentation of the rest of the string. I am going to update the spec and we'll see what others think.

Reply:
That is an excellent question. Which shows that the Java people already thought about this quite well. In case I want, or do not care about the final newline I can also write:
```
String s2 = """
    Text that I do not want indented
    But I do want it indented in the source file
    Indentation in the source makes it easier to read
    """ 

 # Looks nice and properly strips indentation.
```
You have convinced me that we should keep the java spec. It is simple and there is no ambiguity. I see now that suggestions like the above only serve to introduce corner cases which are ambiguous, and that is not desirable.

Reply:
I def stumbled with this once or twice writing the test cases. If we keep the last-line whitespace significant then maybe we also add a linter warning when it's determining the common leading whitespace. (I'm not vetoing since I think "what java does" is a reasonable heuristic, but I also think @rhpvorderman example is persuasive and now backed up by practical experience)